### ISSUE TITLE
Modify TopicFlow Data and Code to Understand how New Data can be Displayed

### ISSUE THREAD

**Issue by estepona** [sailuh/topicflow#2]
I took a look at the LDA algorithm and explored a LDA-based topic modeling library called [gensim](https://radimrehurek.com/gensim/) to see what the definition of the Topics shown in Topic Flow was, however changing the existing data example did not modify the display of topic flow. 

I have an understanding of how each file works and I thought I found the links, but I couldn't change the "data" folder at all. For example, copying a data source "CHI", renaming it, renaming all the inside functions, then adding the renamed functions and ids in "controller.js" and "index.html" would not make the tool load the new data, not to mention generating new data and visualize it. There seem to be hidden functions somewhere controlling the display.

## Proposed next steps:

- Continue to modify TopicFlow code/data and to try to visualize new data
- Refresh my knowledge on d3 and get familiar with sankey.js, a library utilized by TopicFlow.

## Question:
- Based on my understanding, TopicFlow doesn't fit raw text data into their modified LDA model. Am I responsible for implementing the model, or Shashank? Please correct me if I'm wrong.

**Comment by carlosparadis** [comment]
## Proposal

Your proposal looks good. To be even more specific:

 -  I would suggest you try a few other modifications along the 3 files to see if it doesn't reflect any change when you run the application. For example, in one of them, try to edit the content of the Tweet and see if on the right panel when searching for the word (try some crazy word) it doesn't show there.
- Modify the similarity score of a given bin and see if the edge doesn't grow much biggger between the pair of months, or try to use the slider on the left panel to filter lower values so you can see if it is not there.
- See what the other file contains and edit accordingly.

Beware that some of the files have consistency in how they are called. For example, one file may be called bin 0, and in another file there may be a similarity 0_3. When modifying, make sure the id or anything you are changing is not breaking this consistency, as this my try errors which are then like to be invisible to you (thanks, javascript). Also, keep an eye on the terminal when you run it, to see if no other errors are shown. 

## Question

Neither of you. I already have that done, please see [this issue](https://github.com/sailuh/perceive/issues/54) and in particular (the notebook) [in the commit of that pull request](https://github.com/sailuh/perceive/blob/c356393d9fccfe969a474a52198b16e192bda906/Notebooks/LDA/lda_batch.Rmd). What will happen is, I will send you some **Topic-Term-Matrix** and **Document-Topic-Matrix**, along with **similarity scores matrix** them, and you will need to create a few python or R routines to reshape it on the format understood by TopicFlow. 

You should only spend the time to understand LDA to the extent of you know what those 3 matrices are, and how they were used on the TopicFlow paper to be displayed by their tool, and how it is shown on the data models (which is why I had you start there). Don't worry about spending time on trying to understand how libraries are doing (or even how I am doing it). The main part right now is getting TopicFlow to be able to reflect data changes and eventually new data.

**Comment by estepona** [comment]
Finally something works!  

I examined several files and now I'm able to modify the _TopicSimilarity.json_ with whatever source-target pairs or similarity scores I want and reflect the changes in the application. I ran 3 tests overall and all 3 were successful:
- **test** (test0): I copied and pasted the "ModernFamily" data folder and named it "test", then added and changed several function names in related scripts. It displayed exactly the same as the "ModernFamily" in TopicFlow, but notice the title is "TEST". Here's the graph: ![test0](https://raw.githubusercontent.com/estepona/PERCEIVE-freddie/master/test/TopicSimilarity/screenshots/test.png)
- **test1**: based on test (test0), I changed all 292 similarity scores. In the graph below, by looking closely, you will find that the same lines differ in their thicknesses. ![test1](https://raw.githubusercontent.com/estepona/PERCEIVE-freddie/master/test/TopicSimilarity/screenshots/test1.png)
- **test2**: this time I changed the number of links, the source and target of every link, and assigned each one a new similarity score. All the numbers involved are random chosen numbers based on several rules (e.g. each source and target pair has to happen between two adjacent layers, and the consistency problem you mentioned earlier). This new graph looks like: ![test2](https://raw.githubusercontent.com/estepona/PERCEIVE-freddie/master/test/TopicSimilarity/screenshots/test2.png) The links (189 in total) are entirely new to the previous two tests.

The scripts for data transformation can be accessed [here](https://github.com/estepona/PERCEIVE-freddie/blob/master/TopicFlow_topicSimilarityTest_Zhang_09-14.ipynb). It's a bit long. I included all the steps. It turns out that I failed last week because I missed several '\n's and a semi-colon.

So far I haven't made any changes to the **Bins.json** file. I plan to run some tests this week. It will take longer than this one. But I think I'm getting a hang of it. 

What do you think?

**Comment by carlosparadis** [comment]
Could you elaborate a bit on the 

> I missed several '\n\'s and a semicolon 

Why missing the line breaks affect the data changes being reflected on the actual evaluation?

> changed several function names in related scripts

Did you have to change the actual code for the visualization to be changed, or just the data?

Let's discuss this before you do more changes to the data by hand. The main goal is that you understand the overall structure and if overall changing the data you can get a new visualization (which was what I was expecting), if so, then we should be working now on the Notebook to reshape the data the way PERCEIVE has, and have it being reformatted to the way Topic Flow expects instead. I am sure as you go along trying to do that task, you will end up having to learn the structure as well either way.

**Comment by estepona** [comment]
Sure Carlos. For the first question:
> I missed several '\n's and a semicolon  

When reading *Bins.json*, *TopicSimilarity.json*, and *Tweet.json* in Python, I have to read it as a long string at first because the scripts are more formatted like .js files rather than .json files. They have function names and JavaScript syntax. And reading the whole thing as a syntax gives me several '\n's as line breaks (which you  can't see if you open it in text editors). By splitting the long string with '\n' I could locate the actual json-structured data and use the json library to modify existing data or generate new data. I didn't realize it last week so I ran into some formatting problems.

For the second question:
> changed several function names in related scripts.

Yes, both the data and actual code for visualization have to be changed in order to reflect any changes in the application. For example, in order to display a new data set called "test", I did the following changes:
- create a folder called "test" under "data" folder, have the three .json files under "test" folder: *Bins.json*, *TopicSimilarity.json*, and *Tweet.json*;
- for each of the three files, change the function name on the first line to *_test. For example, *populate_bins_test* instead of *populate_bins_ModernFamily* or *populate_bins_CHI*;
- add a new "test" block in *controller.js* (from line 965) under "scripts" folder, as well as a new "test" idToName pair (line 961) so that the application knows there is a new data set called "test" and if this data set is selected, the three functions (*populate_bins_test*, *populate_similarity_test*, *populate_tweets_test*) will be called to display the actual data;
- add a new "test" block in *index.html* (from line 22) , as well as a new "test" selector (line 61) so that the application gives user a choice to select the new data set and the application knows where to find the data. 
- and most importantly, format the data carefully so the application can read it.

You are right about the structure thing. I took much longer time understanding it than writing scripts to format data. I think I'm getting more familiar of the structure of TopicFlow, how the application works and how every file works.

**Comment by carlosparadis** [comment]
>> I missed several '\n's and a semicolon

> When reading Bins.json, TopicSimilarity.json, and Tweet.json in Python, I have to read it as a long string at first because the scripts are more formatted like .js files rather than .json files. They have function names and JavaScript syntax. And reading the whole thing as a syntax gives me several '\n's as line breaks (which you can't see if you open it in text editors). By splitting the long string with '\n' I could locate the actual json-structured data and use the json library to modify existing data or generate new data. I didn't realize it last week so I ran into some formatting problems.

Nice catch. I wonder why they would do such a thing. Now we know how to get the actual data out of it. So, the issue you couldn't see changes was because you were having data errors for missing the `\n`, is that correct?

How is their code reading it in, though? Are they expecting a string as input or the actual .json format? We ought to fix that then create a transformation function that creates some weird format.

> Yes, both the data and actual code for visualization have to be changed in order to reflect any changes in the application.

Interesting, and unfortunate. Sounds like a lot of hard-coded parts for something to be used practically in any dataset. I see now why there is no data transformation pipeline. So, how is it that for the 4 different datasets it works? There is a code clone of the functions you listed above that you need to edit, or is it parameterized? 

Maybe you ought to spend more time dissecting the code to find how much effort would need for you to be able to completely take these hard-coded changes to add different datasets. We will want to try many different formats. Having to change code functions for each one of them is not feasible. As such, do let me know what is your level of comfort in javascript in being able to modify this code to no longer require hard-coded changes for every new dataset (as one would it expect to be). 

A great example of this is LDAvis (an R Package library). Provided the output of the LDA algorithm in a certain format, I don't need to change its code to see plots. 

Great job on finding this out. I am glad I matched you to this problem :-)

**Comment by estepona** [comment]
For the weird .json issue, I just performed a little experiment: I changed the file extension names of the three files under "test" data set folder from:
-  _Bins.json_, _TopicSimilarity.json_, _Tweet.json_   

to:
-  _Bins.js_, _TopicSimilarity.js_, _Tweet.js_

then changed the these file names to be called in _index.html_, and nothing crashes. TopicFlow loads the data just as the same. Even better, sublime now recognizes the JavaScript syntax and give them colors.

I feel like being played by a simple bug, and people who created TopicFlow must know that .json files should only store data. But I might be wrong, they might ran into some issues. Of course, I could use the same script to transform data but it seems weird to write JavaScript codes in Python. 

As far as I can see right now, a simple yet effective solution to display new data sets can be just giving it a standard name, e.g. "new data". All we need to change are two or three lines of code that relates to the name "new data". Function names and other parts can remain what they were. The one but probably the biggest downside is that we can only display one new data set at a time. Like we only possess one cup to fill drinks. Though in our case we can store the old drinks somewhere else. 

With my previous experience writing a JavaScript-based dashboard that loads .json data, I'm sure there is a way to have JavaScript functions only in .js files and only data in .json files. I may spend a week or two tweaking the codes and consider some additional functionalities implied by your thoughts, e.g. a drop-down menu to select data sets when the selections grow in size.

I'll look up LDAvis as well. Looking forward to your thoughts!

**Comment by carlosparadis** [comment]
We would like to move our data to a database at some point, in such a way the script you create reads from a set of tables in the database and store in another set of tables in the database. Someone interested in seeing some data range in Topic Flow would merely specify the time range and the data would run LDA to create the topics, and then pass the tables to Topic Flow. 

I would like to stick to the usual way to plot things, as I pointed to you on LDAVis: We provide the input data, we don't touch any code, and the plot comes out. I can't imagine javascript will require such weird workaround for every new dataset we generate or some convention we must call file something to run it. We should just need to specify where each of the required files is (provided they are in the required format to load in topic flow) for it to run. 

Let me know if this makes sense.

**Comment by estepona** [comment]
Thanks for the feedback. Yes it makes sense. TopicFlow should be as simple as LDAvis. I'll let you know my updates on TopicFlow. 

One question that I'm curious: do I need to touch database or PHP sometime later?

**Comment by carlosparadis** [comment]
Databases, maybe. PHP no. Why do you ask?

**Comment by estepona** [comment]
If I should load data from a database, PHP could be a choice. I'm not familiar with other options.

**Comment by carlosparadis** [comment]
@estepona There are other (way simpler) alternatives, but don't worry about it for now. We first need to ensure we can get a plot given a set of .csv files. Loading from elsewhere can come afterward. 

The expectation is that by the end of this issue you can send a commit with item 1 and 2 (not 2.1) of the list below:

 1. Change the relevant parts in the javascript code so that given some input file in some standard it can plot it, much in the spirit LDAvis does, and  
 2. A python script that takes in the (relevant) outputs of our existing PERCEIVE pipeline, translate into the standard format above, much in the spirit of what I did to get LDAvis to plot the data we are generating. 
   2.1 A variant of this python script that gets the same data from a database sitting on our server.
3. A python notebook using the tool and script you created from various inputs I provide you to see on TopicFlow and a thorough narrative of your interpretation of your results. 

At the point, you create items 1 and 2 (not 2.1) I will create a separate issue for the remaining items. 

This (the entire list items 1 to 3) will be the checkpoint halfway about a month from now.

The rest of the semester will be on something else :-) 

Regards,

**Comment by estepona** [comment]
Thank you for the guideline. It's very clear. In specifying the outputs of existing PERCEIVE pipeline, do you mean the csv files in https://github.com/sailuh/perceive/tree/master/Analysis/Similarity/output?

The career fairs got my head last week. I look forward to a productive afternoon on TopicFlow this week!

**Comment by carlosparadis** [comment]
The pipeline Pal created is retired, the current one is [here](https://github.com/sailuh/perceive/blob/master/Notebooks/LDA/lda_batch.Rmd).

However, I didn't create a Notebook yet showing with some data. I suggest you do **not** look at the code and just focus on items 1 and 2. I do not want you to get biased into modifying Topic Flow interface to accommodate our data and the functions (which are [here](https://github.com/sailuh/perceive/blob/master/Notebooks/LDA/lda_functions.R)) in the same folder. 

For one, because it is still pending a good number of (self) code reviews. Secondly, we should defer the strong transformations to this "mapper script" rather than in the code interface.

**Comment by estepona** [comment]
Hi Carlos, here's the follow up on the issue. I modified the TopicFlow in ways that it only load and display one data set that user specifies. Please download this version [here](https://github.com/estepona/PERCEIVE-freddie/tree/master/test/test%20-%20Copy2%20-%209-30/topicflow) and play with the little demo:
- notice that under "data" folder, there are three directories "archive", "test", and "TF-data". The python program *data_transformation.py* lets you specify the directory that contains three standard .json files to be transformed. Running the program will move the existing .js files from "test" folder to a new directory (e.g data_2017-10-1-9-23) under "archive" folder, and then transform three standard .json files under "TF-data" to TopicFlow readable .js files and write to "test" folder;
- then you can establish a localhost to see the visualization of the .json files;
- the three .json files are actually the "Sandy" data set in the original TopicFlow;
- basically the "archive" part can be removed if not needed.

So far this is what I can come up with. Of more than ten hours I spent on different methods to directly read .json files and display, I failed. The functions in TopicFlow are so intertwined and strict to the input format, changing one part inevitably causes problems in other parts. So I ended up either disabled or removed some sections. 

I didn't look at the code you mentioned in the above block. So long as the input format follows the rule, this approach would work fine. Right now the python program asks for the folder location, but if the ultimate pipeline has a fixed location for the folder, this part can be easily removed.

**Comment by carlosparadis** [comment]
How about this:

- Start from the original code that sits on the perceive repository as to avoid unnecessary changes
- Create a main.py file, which to run takes as input 3 paths: the topic-term-matrix, the document-topic matrix, and the document-frequency-matrix (I believe these are all the necessary inputs). No assumptions on these files other than their structure.
- The main.py is responsible to modify the files as this javascript is coded. Create one function for every logical step of reshape, and document it accordingly as to explain why it needs to be reshaped like so (e.g. you can refer the function that needs the reshape in the javascript side). 
- Follow Google Python Style for function documentation and formatting.
- At some point, your `main.py` will evoke `python3 -m http.server 8000`. Try to pass the reshaped files **in memory** rather than saving and moving files around. 

That way, we keep away from touching the original javascript code, the changes are left in Python to adapt to it, and we can more easily reshape and build things around as needed. 

Let me know what you think, but saving dozens of intermediate files and moving them around in folders is the road to madness (trust me, I have a system in a different department and project that does that and is a nightmare).

**Comment by estepona** [comment]
Sounds good! I'll look into your LDA code and documentation, and ask you if I need more elaboration, hopefully get some parts done by this Friday night. 

I drew a rough [graph](https://raw.githubusercontent.com/estepona/PERCEIVE-freddie/master/whole%20workflow.png) the other day when trying to clear my mind. Could you take a look? Just want to make sure I'm not deviating my effort.

**Comment by carlosparadis** [comment]
Don't worry so much about the LDA code, but rather worry about the data output of it that I will provide. That code is at best on a beta level that I needed to get out so other things on our pipeline wouldn't choke. I will provide you a link on slack briefly about it with some output data you can use to model everything else. 

The diagram is good, can you fix the size of the rectangle where the text is leaking out, and color the blue text from "My Part" to "Topic Flow" and attach the xml and another .png here? I think it would be a good diagram for me to add to the README.md of this project afterward. 

There should be 3 tables that I need to provide you, but the figure the way it is is fine. I will let you know once I post it there.

**Comment by carlosparadis** [comment]
@estepona ok data link was shared on slack. Let me know if you need any clarifications.

**Comment by estepona** [comment]
Hi Carlos, I'm working on the data you provided right now. Could you give some clarification on the _topic_flow.csv_? I don't quite understand the structure of this file. Why 45 rows?

**Comment by carlosparadis** [comment]
Ah..right I forgot this. The reason is readability. As you know, the original code doesn't have a pipeline, and since we needed to see something before Topic Flow was up, this table only shows the highest probability to the next topic, rather than create 10^12 rows which no one would be able to read.

I need to create a second TopicFlow function that generates all of them for you to plug into Topic Flow. Do you think you could work with this data for now? The format will be the same, it will just have way more rows. 

The other tables should be the exact same and contain all the information you need, it is just `topic_flow.csv` that needs to have the other combinations added. 

You can try to create a "quick and dirty" mockup of the input data if you like by modifying the table I sent you and adding the missing rows with 0 probability if it is a must. 

I may be able to get this out quicker than what I think today. So try to start by reshaping the other tables, and leave `topic_flow.csv` as the last part.

**Comment by estepona** [comment]
Thanks for the clarification. Since the structure of the data you provided and the structure TopicFlow needs are different, I need more information to properly map the two:
- for the data you provided, does the 10 topics cover all year?
- if it does, am I correct that there should be a maximum of 10 topics * 10 topics * 11 intervals = 1100 possible combinations overall?
- in terms of the topic model input, TopicFlow needs both topic-document probabilities and document-topic probabilities which are not included in the standard data files. Just for a quick lookup, you have specified the data input [here](https://raw.githubusercontent.com/sailuh/topicflow/master/data_model/bins_model.png).
- another needed input are the .txt files of all months. TopicFlow is so designed for tweets that I think I have to transform the .txt files into "tweet format" so TopicFlow could work. 

For the diagram part, please see the modified .png and .xml files [here](https://github.com/estepona/PERCEIVE-freddie/tree/master/work%20flow%20diagram).

**Comment by carlosparadis** [comment]
It is probably faster if I go over the data with you on 10 minutes on a call with screen share. Are you able to do now? If not you will have to go with what I provided until at least W. 

I don't know what you mean by Tweet format. At the end of the day, a tweet of a research article is a list of words where the order of them matters. 

I provided you with a Topic-Term Matrix of 10 topics for every month. LDA is run on every month and create 10 topics for every month. That is why you see every table named after a month. Our Topic Flow implementation then uses, mostly in the same fashion as the one described in the paper, similarity on pairs of topics to decide on the highest. 

The one extra thing you may need is the folder with the actual parsed documents from the mailing list. That I can provide you today on Slack since it is already on Mega.

**Comment by estepona** [comment]
I went to bed last night after the reply :P. I still see some differences between the two applications that don't quite match. A call would be faster to go over this! Are you available this afternoon (Hawaii Time) or tomorrow?

**Comment by estepona** [comment]
Hi, Carlos. Here's the update of my part. I did a test on 2012 data. Most of it works, but some parts not. A link to the notebook is [here](https://github.com/estepona/PERCEIVE-freddie/blob/master/10-7%20LDA%20data%20transformation.ipynb). The notebook is not styled yet, but I have put many comments so that you can get the idea the way I'm approaching. Running TopicFlow with generated three files give me this:
![image](https://raw.githubusercontent.com/estepona/PERCEIVE-freddie/master/test/TopicSimilarity/screenshots/Oct_15_TopicFlow_result.png)
You can see that it's ready to work, but some parts crashed. When trying to debug, the most bothering one is a "style" error that has happened to me many times during the previous works on TopicFlow, and the original code regarding this part is very hard to read. I'm kind of stuck here.  I'll try more debugging tomorrow.

**Comment by carlosparadis** [comment]
Hi, Zhang, thanks for the update.


The image is grayed out, can you try to retake the screenshot?


I am not sure what you mean by:

> but I have put many comments so that you can get the idea the way I'm approaching.

I need some narrative in the notebook to try to make sense of it. Right now it feels like the comments were made as reminders rather explain me something. 

Also,

> The notebook is not styled yet, but I have put many comments so that you can get the idea the way I'm approaching

Rather than have me try to "get the idea", can you tell me here what the idea is?

**Comment by estepona** [comment]
Sorry about the ambiguity. First of all, the screenshot is grayed because it can't configure some parts out correctly, not until everything is fine with the codes. 

The idea is that every chunk of code aims to transform data into one of three .js files. To do this, I either put all the relevant data into a pandas.Dataframe or a big dictionary, then transform into json format, then add JS codes.
